src/group-by.mjs

```diff
@@ -1,39 +1,96 @@
 import { iterableCurry } from './internal/iterable'
+import { Exchange } from './internal/queues'
 
-function * groupBy (key = (k) => k, iterable) {
+const UNIQUE_INITIAL_KEY = {}
+
+function groupBy (getKey = (k) => k, iterable) {
   const iterator = iterable[Symbol.iterator]()
 
+  let itemIndex = 0
+  let iterableCounter = 0
+  let noNewIterables = false
+  const exchange = new Exchange()
+  let done = false
   let currentItem
-  let currentKey, previousKey
-
-  function * group () {
-    while (true) {
-      yield currentItem.value
-      currentItem = iterator.next()
-      if (currentItem.done) return
-      currentKey = key(currentItem.value)
-      if (previousKey !== currentKey) {
-        return
+
+  // fetch new item from Iterator
+  // return the item and advance the
+  // main queue consumer
+  function fetch () {
+    const newItem = iterator.next()
+    if (newItem.done) {
+      done = true
+      return
+    }
+    currentItem = { value: newItem.value, key: getKey(newItem.value, itemIndex++) }
+    exchange.push(currentItem)
+  }
+
+  // close the original iterator if possible
+  function returnIterator () {
+    if (noNewIterables && iterableCounter === 0) {
+      if (typeof iterator.return === 'function') iterator.return()
+    }
+  }
+
+  // generate subgroup where adjacent items have the same key
+  // it picks up new item from the buffer. Every instance has
+  // its own independent consumer
+  function * generateGroup (firstItem, cons) {
+    try {
+      iterableCounter++
+      yield 'ensure finally'
+
+      yield firstItem.value
+
+      while (true) {
+        if (cons.isEmpty()) {
+          fetch()
+          if (done) return
+          continue
+        }
+        const nextItem = cons.shift()
+        if (nextItem.key !== firstItem.key) {
+          return
+        }
+        yield nextItem.value
       }
+    } finally {
+      iterableCounter--
+      returnIterator()
     }
-  };
-
-  try {
-    currentItem = iterator.next()
-
-    while (true) {
-      if (currentItem.done) return
-      currentKey = key(currentItem.value)
-      if (previousKey !== currentKey) {
-        previousKey = currentKey
-        yield [currentKey, group()]
-      } else {
-        currentItem = iterator.next()
+  }
+
+  // generate group returning [key, subgroup]
+  // it always picks up new items from the Iterator (fetch)
+  function * generateGroups () {
+    let currentKey = UNIQUE_INITIAL_KEY
+    try {
+      while (true) {
+        // I need to fetch a new item if currentItem is undefined (first time)
+        // or
+        // currentItem.key === currentKey that means that currentItem
+        // a fetch called in a subgroup returned an item of a different key.
+        if (!currentItem || currentItem.key === currentKey) {
+          fetch()
+        }
+
+        if (done) return
+
+        if (currentItem.key !== currentKey) {
```

Comment:
I would like a better comment here (or nearby). Actually I think a lot of the current comments don't tell much more than the code itself does, but here especially is where I got tripped up.

This line is the core of how/why your implementation works, and I only just fully understood it now. It causes the main loop to have a double purpose (triple?) purpose. It yields groups, the fetch handles one-item lookaheads (to tell if the next item be part of this group), and the combination of those two also serve the purpose of fetching the entire previous group, but only if it has not already been fetched as a result of the code on the other side of the `yield` consuming the result of `generateGroups()`. I understand now that our implementations are effectively equivalent in their primary technique and behavior.

Reply:
You are in the ideal position to suggest the best possible comment here :-)